src/ops/custom.rs

```diff
@@ -1,219 +1,186 @@
-use crate::macros::impl_box_clone;
-use crate::resource::ResourceSet;
-use crate::types::SimpleType;
-use crate::types::{Signature, SignatureDescription};
+use crate::hugr::{HugrMut, HugrView};
+use crate::resource::{OpDef, ResourceId, ResourceSet};
+use crate::types::{type_param::TypeArg, Signature, SignatureDescription};
+use crate::{Hugr, Resource};
 
-/// A wrapped [`CustomOp`] with fast equality checks.
-#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
-pub struct OpaqueOp {
-    /// Operation name, cached for fast equality checks.
-    id: SmolStr,
+use super::tag::OpTag;
+use super::{LeafOp, OpName, OpTrait, OpType};
 
-    /// The custom operation.
-    custom: Box<dyn CustomOp>,
+/// An instantiation of an [`OpDef`] with values for the type arguments
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+#[serde(into = "OpaqueOp", from = "OpaqueOp")]
+pub enum ExternalOp {
+    Resource { def: Arc<OpDef>, args: Vec<TypeArg> },
+    Opaque(OpaqueOp),
 }
 
-impl OpaqueOp {
-    /// Initialize a new named OpaqueOp
-    pub fn new(id: impl Into<SmolStr>, custom: impl CustomOp) -> Self {
-        Self {
-            id: id.into(),
-
-            custom: Box::new(custom),
+impl From<ExternalOp> for OpaqueOp {
+    fn from(value: ExternalOp) -> Self {
+        match value {
+            ExternalOp::Opaque(op) => op,
+            ExternalOp::Resource { def, args } => {
+                // There's no way to report a panic here, but serde requires Into not TryInto. Eeeek!
+                let signature = if def.should_serialize_signature() {
+                    // TODO More issues with the input resource set here!
+                    Some(def.signature(&args, &ResourceSet::default()).unwrap())
+                } else {
+                    None
+                };
+                OpaqueOp {
+                    resource: def.resource.clone(),
+                    op_name: def.name.clone(),
+                    description: def.description.clone(),
+                    args,
+                    signature,
+                }
+            }
         }
     }
+}
 
-    /// The name of the operation, cached for fast equality checks.
-    pub fn name(&self) -> SmolStr {
-        self.id.clone()
+impl From<OpaqueOp> for ExternalOp {
+    fn from(op: OpaqueOp) -> Self {
+        Self::Opaque(op)
     }
 }
 
-impl PartialEq for OpaqueOp {
-    fn eq(&self, other: &Self) -> bool {
-        self.id == other.id
+impl From<ExternalOp> for LeafOp {
+    fn from(value: ExternalOp) -> Self {
+        LeafOp::CustomOp(value)
     }
 }
 
-impl Eq for OpaqueOp {}
-
-impl Deref for OpaqueOp {
-    type Target = dyn CustomOp;
+/// An opaquely-serialized op that refers to an as-yet-unresolved [`OpDef`]
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct OpaqueOp {
+    resource: ResourceId,
+    op_name: SmolStr,
+    #[serde(skip)]
+    description: String, // cache in advance as description() returns &str
```

Comment:
Why should we skip this in serialisation? Wont we wont it to be exposed to/from guppy?

Reply:
Yes, good point. I was thinking that when we turn this back into a ResourceOp we'll lose this (the `OpDef` defines the description method instead). So we could probably benefit from some mechanism for serializing different parts at different times, but I've not done that here in this "sketch"! However, description now saved and restored.

When we turn back into a ResourceOp we'll just discard the serialized description - the only other thing we could really do would be to raise an error if they were different, but I haven't bothered to do that here.